### PR TITLE
Add nextcloud host to capabilities response

### PR DESCRIPTION
**Pull request by kesselb**
We have a report (https://github.com/nextcloud/server/issues/15688) that some requests by the desktop client trigger fail2ban. 
The reason seems a htaccess rule/redirect: https://github.com/nextcloud/server/issues/15688#issuecomment-707973799.

It seems the desktop client needs this information for the copy internal link feature: https://github.com/nextcloud/server/issues/15688#issuecomment-724101752 / https://github.com/nextcloud/desktop/pull/2619.

Desktop client does the request for capabilities and ocs config in parallel. I guess it makes sense to provide the host information also within the capabilities request to avoid the second request.

### CODE REVIEW

File: tests/Core/Controller/OCSControllerTest.php [hunk 1]
```diff
@@ -88,6 +88,10 @@ public function testGetCapabilities() {
 		$this->userSession->expects($this->once())
 			->method('isLoggedIn')
 			->willReturn(true);
+		$this->request
+			->method('getServerHost')
+			->willReturn('awesomehost.io');
```
Comment (nickvergessen) [inline tests/Core/Controller/OCSControllerTest.php:93]:
```suggestion
			->willReturn('example.tld');
```
or a nextcloud domain

**Review by nickvergessen — CHANGES_REQUESTED** [review]
So the client makes a valid address to get the capabilities and we then return the same valid host?

**Comment by kesselb** [comment]
> So the client makes a valid address to get the capabilities and we then return the same valid host?

Yes ;) That's what the desktop client does at the moment. Sent a request to ocs/v1.php/config and set host as setUserVisibleHost. I think providing host via capabilities was something we discussed with Kevin. If you suggest to extract the user visible host from the given url the client already has - works for me as well ;)

**Comment by kesselb** [comment]
```

    /** If url to use for any user-visible urls.
     *
     * If the server configures overwritehost this can be different from
     * the connection url in _url. We retrieve the visible host through
     * the ocs/v1.php/config endpoint in ConnectionValidator.
     */
    QUrl _userVisibleUrl;
```

https://github.com/nextcloud/desktop/blob/7e5f080e58f5c2074b9cfd417f1621204bd6f96c/src/libsync/account.h#L310-L316. Does that make sense?

**Comment by nickvergessen** [comment]
The documentation sounds like server config `overwrite.cli.url` should be used.

**Comment by kesselb** [comment]
> The documentation sounds like server config `overwrite.cli.url` should be used.

`overwrite.cli.url` is static. getServerHost is overwritehost or a trusted_domain. 

https://github.com/nextcloud/server/blob/05742c083c039561841d258a69fd47dd84debfd6/apps/files_sharing/src/components/SharingEntryInternal.vue#L50-L56

https://github.com/nextcloud/android/blob/247d422b3748f0c2b6ea270bdf3613dba0abe6cc/src/main/java/com/owncloud/android/ui/fragment/FileDetailSharingFragment.java#L267-L269

Should also work for the desktop client. Web and Android are not reading the host from ocs/v1.php/config.

**Comment by tobiasKaminsky** [comment]
On Android we do:
- let user enter server address
- do a server info check http(s)://$serverAddress/status.php
- return those info *and* use server address from above

So basically we rely on the address the user entered when first setting up the account.
And this is then also used to create internal share links: $serverAddress + /f/$id

**Comment by kesselb** [comment]
Closing in favor of https://github.com/nextcloud/desktop/pull/3552.

**Pull request closed without merging**